Reject master keys shorter than 16 bytes in `production` (#209)
meilisearch/specifications: text/0119-instance-options.md

You must supply an alphanumeric string when using this option.

If no master key is provided in a `development` environment, all routes will be unprotected and publicly accessible.

##### 3.3.4.1. Error and Warning messages

Providing a master key of at least 16 bytes is mandatory when `--env` is set to `production`; if none is given, Meilisearch will throw an error and refuse to launch.

```
Error: In production mode, you must provide a master key to secure your instance. It can be specified via the MEILI_MASTER_KEY environment variable or the --master-key launch option.

We generated a secure master key for you (you can safely copy this token):

>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

Providing a master key of at least 16 bytes is mandatory when `--env` is set to `production`; if it is given but too short then, Meilisearch will throw an error and refuse to launch.

```

> Error: In production mode, the master key must be of at least 16 bytes, but the provided key is only `:numBytes` bytes long

Reviewer: ```suggestion
Error: In production mode, the master key must be of at least 16 bytes, but the provided key is only `:numBytes` bytes long.
```

We generated a secure master key for you (you can safely copy this token):

>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

Providing a master key of less than 16 bytes when `--env` is set to `development` displays a warning message.

```

> The provided master key is too short (< 16 bytes)
> A master key of at least 16 bytes will be required when switching to the production environment.

Reviewer: ```suggestion
The provided master key is too short (< 16 bytes) and thus considered unsafe for production use.
A master key of at least 16 bytes will be required when switching to the production environment.

We generated a secure master key for you (you can safely copy this token):

>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

Author: Concern: users could believe that Meilisearch was started with the generated key rather than the one they provided.

This condition is unique: Meilisearch accepts to start with the provided master key that is too short, but still generates a different one. A possible interpretation could be that Meilisearch used the generated key instead of the provided one.

In a test with a fellow team member, this was their interpretation.

I feel we should specify somehow that the generated key is a suggestion to use for future launch, and maybe even that meilisearch started with the provided key.

Reviewer: Proposal 1: Not showing the master key currently being used in the message.

```
Meilisearch started with a master key considered unsafe for production use.
	
A master key of at least 16 bytes will be required when switching to the production environment.
	
We generated a secure master key for you if you want to change it now (you can safely copy this token):
	
>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

Proposal 2: Showing the master key currently being used in the message.

```
Meilisearch started with the following master key:

>> `:usedMasterKey` <<
	
The provided master key is too short (< 16 bytes) and should be considered unsafe. A master key of at least 16 bytes will be required when switching to the production environment.
	
We generated a secure master key for you if you want to change it now (you can safely copy this token):
	
>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

Is it better?

Author: I think proposal 2 is a no-no, we should never print a master key that was *user-provided* to stdout.

I like proposal 1, maybe even stressing it further:

```
Meilisearch started with a master key considered unsafe for production use.
	
A master key of at least 16 bytes will be required when switching to the production environment.
	
We generated a new secure master key for you (you can safely copy this token):
	
>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<

Restart Meilisearch after executing the command above to use this new and secure master key.
```

What do you think?

Reviewer: > I think proposal 2 is a no-no, we should never print a master key that was user-provided to stdout.

Agree!

I like your proposal. 🪄 Should we add the final line to every message?

Author: > Should we add the final line to every message?

Hmm I'm not sure, especially in the error cases, since we already stop meilisearch for the user

Reviewer 2: > Restart Meilisearch after executing the command above to use this new and secure master key.

I completely understand the message's purpose, but if the user does it literally, it may not work when using Docker.

Because if you run the `export` command, the ENV var will be applied to the host, and the user will still have to pass it through the docker command.
So it does not work if the user "just restart Meilisearch".

```

Not providing a master key when `--env` is set to `development` displays a warning message.

```
No master key found; The server will accept unidentified requests. If you need some protection in development mode, please export a key:

export MEILI_MASTER_KEY=`:suggestedMasterKey`

> A master key of at least 16 bytes will be required when switching to the production environment.

Reviewer: No master key found; The server will accept unidentified requests.

A master key of at least 16 bytes will be required when switching to the production environment.

If you need some protection in development mode, we generated a secure master key for you (you can safely copy this token):

>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<

> No master key found; The server will accept unidentified requests. If you need some protection in development mode, please export a key:
> 
> export MEILI_MASTER_KEY=`:suggestedMasterKey`
> 
> A master key of at least 16 bytes will be required when switching to the production environment.

Reviewer: ```suggestion
No master key found; The server will accept unidentified requests.

A master key of at least 16 bytes will be required when switching to the production environment.

If you need some protection in development mode, we generated a secure master key for you (you can safely copy this token):

>> export MEILI_MASTER_KEY=`:suggestedMasterKey` <<
```

```
Error: You must provide a master key to secure your instance in a production environment. It can be specified via the MEILI_MASTER_KEY environment variable or the --master-key launch option.

We generated a secure master key for you (you can safely use this token):

>> --master-key `:suggestedMasterKey` <<
```

Providing a master key of at least 16 bytes is mandatory when `--env` is set to `production`; if it is given but too short then, Meilisearch will throw an error and refuse to launch.

```
Error: The master key must be at least 16 bytes in a production environment. The provided key is only `:numBytes` bytes.

We generated a secure master key for you (you can safely use this token):

>> --master-key `:suggestedMasterKey` <<
```

Providing a master key of less than 16 bytes when `--env` is set to `development` displays a warning message.

```
Meilisearch started with a master key considered unsafe for use in a production environment.

A master key of at least 16 bytes will be required when switching to a production environment.

We generated a new secure master key for you (you can safely use this token):

>> --master-key `:suggestedMasterKey` <<

> Restart Meilisearch after executing the command above to use this new and secure master key.

Author: Late change notice ...

since this is an additional flag and not a command to execute this sentence should change.

```suggestion
Restart Meilisearch with the argument above to use this new and secure master key.
```

```

Not providing a master key when `--env` is set to `development` displays a warning message.

```

> No master key found; The server will accept unidentified requests.

Reviewer 3: ```suggestion
No master key was found. The server will accept unidentified requests.
```

> A master key of at least 16 bytes will be required when switching to the production environment.

Reviewer 3: ```suggestion
A master key of at least 16 bytes will be required when switching to production mode.
```

> If you need some protection in development mode, we generated a secure master key for you (you can safely copy this token):

Reviewer 3: ```suggestion
If you need protection in a development environment, we generated a secure master key for you (you can safely copy this token):
```

>> --master-key `:suggestedMasterKey` <<

> Restart Meilisearch after executing the command above to use this new and secure master key.

Author: ```suggestion
Restart Meilisearch with the argument above to use this new and secure master key.
```
